modules/core/src/test/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerClientReconnectAfterClusterRestartTest.java

```diff
@@ -0,0 +1,129 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.datastreamer;
+
+import javax.cache.CacheException;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.IgniteClientDisconnectedException;
+import org.apache.ignite.IgniteDataStreamer;
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.spi.discovery.tcp.TcpDiscoverySpi;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.vm.TcpDiscoveryVmIpFinder;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+
+/**
+ * Tests DataStreamer reconnect behaviour when client nodes arrives at the same or different topVer than it left.
+ */
+public class DataStreamerClientReconnectAfterClusterRestartTest extends GridCommonAbstractTest {
+    /** */
+    public static final TcpDiscoveryVmIpFinder IP_FINDER = new TcpDiscoveryVmIpFinder(true);
+
+    /** */
+    private boolean clientMode;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        ((TcpDiscoverySpi)cfg.getDiscoverySpi()).setIpFinder(IP_FINDER);
+
+        cfg.setCacheConfiguration(new CacheConfiguration<>("test"));
+
+        cfg.setClientMode(clientMode);
+
+        return cfg;
+    }
+
+    /** */
+    public void testOneClient() throws Exception {
+        clusterRestart(false, false);
+    }
+
+    /** */
+    public void testOneClientAllowOverwrite() throws Exception {
+        clusterRestart(false, true);
+    }
+
+    /** */
+    public void testTwoClients() throws Exception {
+        clusterRestart(true, false);
+    }
+
+    /** */
+    public void testTwoClientsAllowOverwrite() throws Exception {
+        clusterRestart(true, true);
+    }
+
+    /** */
+    private void clusterRestart(boolean withAnotherClient, boolean allowOverwrite) throws Exception {
+        try {
+            startGrid(0);
+
+            clientMode = true;
+
+            Ignite client = startGrid(1);
+
+            if (withAnotherClient) {
+                startGrid(2);
+
+                stopGrid(2);
+            }
+
+            clientMode = false;
+
+            try (IgniteDataStreamer<String, String> streamer = client.dataStreamer("test")) {
+                streamer.allowOverwrite(allowOverwrite);
+
+                streamer.addData("k1", "v1");
+            }
+
+            stopGrid(0);
+
+            U.sleep(2_000);
```

Comment:
I suggest getting rid of Thread.sleep(). It does not look reliable.
Perhaps, Ignite events will do the required trick:

`
            final CountDownLatch disconnect = new CountDownLatch(1);
            final CountDownLatch reconnect = new CountDownLatch(1);

            Ignite client = startGrid(1);
            client.events().localListen(new IgnitePredicate<Event>() {
                @Override public boolean apply(Event event) {
                    reconnect.countDown();

                    return false;
                }
            }, EventType.EVT_CLIENT_NODE_RECONNECTED);
            client.events().localListen(new IgnitePredicate<Event>() {
                @Override public boolean apply(Event event) {
                    disconnect.countDown();

                    return false;
                }
            }, EventType.EVT_CLIENT_NODE_DISCONNECTED);

    ...

            stopGrid(0);

            disconnect.await();

            startGrid(0);

            reconnect.await();

    ...
`

WDYT?

Reply:
I have implemented that, but test now runs slower :( 8s vs 5s for individual tests